File: python/unbabel.py
#
# unbabel.py
#

import re
import string
import sexp
import os

class Signature:

  """ make a instance of Signature
      >>> sig = Signature("# fib :: Int -> Int", "#")
      >>> sig.funcname
      'fib'
      >>> sig.rettype
      'Int'
      >>> sig.argtypes
      ['Int']
  """
  def __init__(self, src, comment):
    self.parse(src, comment)

  RE_SIGNATURE = "\s*(\S+)\s*::\s*((?:\S+\s*->\s*)+)(\S+)\s*"
  
  def parse(self, src, comment):
    rexp = re.compile(comment + self.RE_SIGNATURE, re.MULTILINE)
    match = rexp.search(src)
    if match:
      (self.funcname, argtypes, self.rettype) = match.groups()
      self.argtypes = map(string.strip, argtypes.split("->")[0:-1])
    else:
      raise RuntimeError("signature not found in script")

# langdef

class LangDef:

  """ make a instance of LangDef
      >>> l = LangDef("../templates/ruby.unb")
      >>> (l.name, l.comment, l.command, l.argtype)
      ('Ruby', '#', 'ruby', 'square')
      >>> l.stub.split("\\n")[0]
      '<FUNCTION_DEFINITION>'
      >>> l.stub.split("\\n")[-3]  # 3..('A`)
      'print Babel.to_sexp(<FUNCTION_NAME>(<ARGUMENTS>))'
  """
  def __init__(self, path):
    f = open(path, "r")
    s = f.read()
    f.close()
    self.parse(s)

  def parse(self, s):
    in_stub = False
    stub = ""
    for line in s.split("\n"):
      if in_stub:
        stub += line + "\n"
      elif line == "stub:":
        in_stub = True
      else:
        (key, value) = line.split(":")
        if key == "name":
          self.name = value
        elif key == "comment":
          self.comment = value
        elif key == "command":
          self.command = value
        elif key == "argtype":
          self.argtype = value
        else:
          raise RuntimeError("unkown key in langdef: " + key)
    self.stub = stub

# stub

class Stub:

  """ make a stub script
      >>> foo = \"\"\"
      ... # test
      ... # foo :: [[Int]] -> Int
      ... foobar
      ... \"\"\"
      >>> stub = Stub(foo, LangDef("../templates/ruby.unb"), [[10, 20], [3, 4]])
      >>> stub.script
      "\\n# test\\n# foo :: [[Int]] -> Int\\nfoobar\\n\\n\\nclass Babel\\n  def self.to_sexp(x)\\n    case x\\n    when Numeric\\n      x.to_s\\n    when String\\n      x.inspect\\n    when Array\\n      '(' + x.map{|item| Babel.to_sexp(item)}.join(' ') + ')'\\n    else\\n      raise\\n    end\\n  end\\nend\\nprint Babel.to_sexp(foo([10, 20], [3, 4]))\\n\\n"
  """
  def __init__(self, src, langdef, args):
    self.langdef = langdef
    self.make(src, args)
  
  def make(self, src, args):
    sig = Signature(src, self.langdef.comment)
    self.script = re.sub("<FUNCTION_DEFINITION>", src,
                  re.sub("<FUNCTION_NAME>", sig.funcname,
                  re.sub("<ARGUMENTS>", self.format_args(args, sig),
                  self.langdef.stub)))
  
  def format_args(self, args, sig):
    formatted = map( lambda arg : self.format_arg(arg, sig, True), args)
    return ", ".join(formatted)
    
  def format_arg(self, arg, sig, outer):
    # TODO: typecheck using sigature
    if isinstance(arg, int):
      return str(arg)
    elif isinstance(arg, str):
      # TODO: I need something like Ruby's String#inspect
      return '"' + re.sub(r"\\", "\\\\", 
                   re.sub(r"\t", "\\t",
                   re.sub(r"\n", "\\n", arg))) + '"'
    elif isinstance(arg, list):
      (pre, between, post) = self.parens(outer)
      formatted = map( lambda x : self.format_arg(x, sig, False), arg)
      return pre + between.join(formatted) + post
    else:
      raise RuntimeError, "unknown type: " + arg

  PARENS = {
    "square": ("[", ", ", "]"),
    "paren":  ("(", ", ", ")"),
    "sexp":   ("(", " ",  ")")
  }
  def parens(self, outer):
    argtype = self.langdef.argtype
    try:
      (pre, between, post) = self.PARENS[argtype]
    except KeyError:
      raise RuntimeError("unknown argtype: " + argtype)
        
    if argtype == "sexp" and outer:
      return ("'" + pre, between, post)
    else:
      return (pre, between, post)

# runner

class Runner: 

  """ run a stub script
      >>> to_i = \"\"\"
      ... # to_i :: String -> Int
      ... def to_i(s); s.to_i; end
      ... \"\"\"
      >>> stub = Stub(to_i, LangDef("../templates/ruby.unb"), ["10"])
      >>> Runner(stub).run()
      10
  """
  def __init__(self, stub):
    self.stub = stub

  def run(self):
    result = self.run_command(self.stub.langdef.command, self.stub.script)
    return sexp.read_sexp(result)
    # TODO: type check by signature

  def run_command(self, command, script):
    (stdin, stdout) = os.popen2(command)
    stdin.write(script)
    stdin.close()
    return stdout.read()

# integration

def init(module, path):
  def makelang(langdef):
    def lang(src):
      """ returns Ruby function
          >>> fib = ruby("# fib :: Int -> Int\\ndef fib(n); 55; end") 
          >>> fib(10)
          55
      """
      def runner(*args):
        stub = Stub(src, langdef, args)
        return Runner(stub).run()
      return runner
    return lang

  for file in os.listdir(path):
    if file.endswith(".unb"):
      langdef = LangDef(os.path.join(path, file))
      setattr(module, file[0:-4], makelang(langdef))

if __name__ == "__main__":
  import doctest
  doctest.testmod()
  

File: ruby/lib/unbabel.rb
require 'tempfile'
require 'fileutils'

class String
  def unindent
    self =~ /^(\s*)/
    indent = $1
    self.lines.map{|line|
      line.sub(indent, "")
    }.join
  end
end

# String#lines for ruby <= 1.8.6
unless "".respond_to?(:lines)
  class String
    def lines
      self.split(/\r?\n/).map{|line| line+"\n"}
    end
  end
end
    

module Unbabel

  # Signeture class: parse signetures
  class Signeture

    REXP_TYPE = /[\[\]\w]+/
    REXP_SIGNETURE = /(#{REXP_TYPE})\s*::\s*(#{REXP_TYPE}(\s*->\s*#{REXP_TYPE})*)\s*$/
    TYPE_TABLE = [
      ["Int", Integer],
      ["String", String],
    ]

    def self.parse(*args)
      new(*args)
    end

    def initialize(str, comment)
      line = find_signeture(str, Regexp.new(comment))
      if line
        parse_signeture(line)
      else
        raise "signeture not found"
      end
    end
    attr_reader :func_name, :return_type, :arg_types

    def find_signeture(str, rexp_comment)
      str.lines.find{|line|
        line =~ /#{rexp_comment}\s*#{REXP_SIGNETURE}/
      }
    end

    def parse_signeture(line)
      line =~ REXP_SIGNETURE
      @func_name = $1.strip
      @arg_types = $2.split(/->/).map{|x| x.strip}
      @return_type = @arg_types.pop
    end
    #TODO 4: multi signetures?

  end

  class Runner

    def self.run(*args)
      new(*args).run
    end

    def initialize(src, args, langdef)
      @src = src
      @args = args
      @langdef = langdef
    end

    def run
      sig = Unbabel::Signeture.parse(@src, @langdef.comment)
      stub = make_stub(@src, @args, @langdef)
      result = run_stub(stub, @langdef)
      return parse_result(result, sig)
    end

    def make_stub(src, args, langdef)
      sig = Unbabel::Signeture.new(src, langdef.comment)
      stub = langdef.stub
      return stub.gsub(/<FUNCTION_DEFINITION>/, src).
                  gsub(/<FUNCTION_NAME>/, sig.func_name).
                  gsub(/<ARGUMENTS>/, langdef.format_args(args))
    end

    def run_stub(script, langdef)
      # make a temporary file
      tempfile = Tempfile.open('babel21')
      tempfile.write(script)
      tempfile.close(false)

      # run it
      if langdef.name == "Java"
        temp_java  = File.expand_path("Unbabel.java", File.dirname(tempfile.path))
        temp_jclass = File.expand_path("Unbabel.class", File.dirname(tempfile.path))
        system "cp #{tempfile.path} #{temp_java}"
        cmd = "#{langdef.command} #{temp_java}"
      else
        cmd = "#{langdef.command} #{tempfile.path}"
      end
      result = `#{cmd}`

      # delete the file
      if langdef.name == "Java"
        FileUtils.rm(temp_java)
        FileUtils.rm(temp_jclass)
      end
      tempfile.open
      tempfile.close(true)

      result
    end

    def parse_result(result, sig)
      case sig.return_type
      when "Int"
        result.to_i
      when "String"
        result
      when "[String]"
        result =~ /\((.*)\)/
        $1.split(/ /).map{|x| 
          x =~ /"(.*)"/
          $1
        }
      when "[Int]"
        result =~ /\((.*)\)/
        $1.split(/ /).map{|x| x.to_i}
      else
        raise "unknown type: #{sig.return_type}"
      end
    end
    #TODO: support [[Int]], [Object], Number, etc.

  end

  LangDef = Struct.new(:name, :comment, :command, :argtype, :stub)

  class LangDef

    def self.load(path)
      defs = {}
      stub = ""

      in_stub = false
      src = File.read(path)
      src.lines.each do |line|
        if in_stub 
          stub << line
        else
          key, value = line.chomp.split(/:/).map{|x| x.strip}
          if key == "stub"
            in_stub = true
          else
            defs[key] = value
          end
        end
      end

      new(defs["name"], defs["comment"], defs["command"],
          defs["argtype"], stub)
    end

    # call format_sexp, format_square, etc.
    def format_args(args)
      method = "format_#{self.argtype}"
      self.__send__(method, args)
    end
    #TODO 5: type check by the signeture

    def format_sexp(args)
      args.map{|x| format_object(x, " ", "'")}.join(' ')
    end

    def format_square(args)
      args.map{|x| x.inspect}.join(', ')
    end

    def format_paren(args)
      args.map{|x| format_object(x, ", ", "")}.join(', ')
    end

    private

    def format_object(x, delimiter, list_prefix)
      case x
      when Array
        list_prefix + "(" + 
          x.map{|item| format_object(item, delimiter, "")}.join(delimiter) + ")"
      when String
        x.inspect
      when Numeric
        x.to_s
      else
        raise "This type of arguments are not supported: #{x.class}: #{x}"
      end
    end

  end

  class Language
    # LANGDEF = LangDef.new(...)   # each subclass has own constant

    def initialize(src)
      @src = src
    end

    def call(*args)
      Unbabel::Runner.run(@src, args, self.class.const_get(:LANGDEF))
    end
    alias [] call

  end

  # TODO: configurable template path
  dir = File.expand_path("../../templates/", File.dirname(__FILE__)) 
  Dir.glob("#{dir}/*.unb").each do |path|
    langdef = Unbabel::LangDef.load(path)

    lang_class = Class.new(Language)
    lang_class.const_set("LANGDEF", langdef)

    Unbabel.const_set(langdef.name, lang_class)
  end

end
